File: python/Longest_Substring_Without_Repeating_Characters/Longest_Substring_Without_Repeating_Characters.py
class Solution(object):
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        freq = {}
        for i in range(256):
            freq[i] = 0
        ls = len(s)
        l=max_len = 0
        r=-1
        while(r+1 < len(s)):
            if freq[ord(s[r+1])] ==0:
                freq[ord(s[r+1])] += 1
                r+=1
            else:
                freq[ord(s[l])] -= 1
                l +=1
            max_len = max(max_len, r - l +1 )
        return max_len

File: cpp/Longest_Substring_Without_Repeating_Characters/Longest_Substring_Without_Repeating_Characters.cpp
/// Source : https://leetcode.com/problems/longest-substring-without-repeating-characters/description/
/// Author : bryce
/// Time   : 2019-10-10

#include <iostream>
#include <string>
#include <cassert>

using namespace std;

// Sliding Window
// Time Complexity: O(len(s))
// Space Complexity: O(len(charset))
class Solution {
public:
    int lengthOfLongestSubstring(string s) {

        int freq[256] = {0};

        int l = 0, r = 0; // sliding window: s[l...r]
        int res = 0;

        while(r < s.size()){
            if( freq[s[r]] == 0 )
            {
               freq[s[r]] ++;
               r++;
             }
            else
            {
               freq[s[l]] --;
               l++;
            }

            res = max(res, r - l);
        }
        return res;
    }
};

class Solution2 {
public:
    int lengthOfLongestSubstring(string s) {
        unordered_map<char, int> hash;
        int res = 0;
        for (int i = 0, j = 0; j < s.size(); j ++ )
        {
            hash[s[j]] ++ ;
            while (hash[s[j]] > 1) hash[s[i ++ ]] -- ;
            res = max(res, j - i + 1);
        }
        return res;
    }
};

int main() {

    cout << Solution().lengthOfLongestSubstring( "abcabcbb" )<<endl; //3
    cout << Solution().lengthOfLongestSubstring( "bbbbb" )<<endl;    //1
    cout << Solution().lengthOfLongestSubstring( "pwwkew" )<<endl;   //3
    cout << Solution().lengthOfLongestSubstring( "c" )<<endl;        //1
    cout << Solution().lengthOfLongestSubstring( "" )<<endl;         //0

    return 0;
}
